src/abstractdataframe/abstractdataframe.jl

```diff
@@ -1535,29 +1546,94 @@ julia> vcat(df1, df3, cols=:intersect)
    6 │     9
 
-julia> vcat(d4, df1)
+julia> vcat(df4, df1)
 3×2 DataFrame
  Row │ A      B
      │ Int64  Int64
 ─────┼──────────────
    1 │     1      1
    2 │     2      2
    3 │     3      3
+
+julia> vcat(df1, df2, df3, df4, cols=:union, source="source")
+10×4 DataFrame
+ Row │ source  A        B        C
+     │ Int64   Int64?   Int64?   Int64?
+─────┼───────────────────────────────────
+   1 │      1        1        1  missing
+   2 │      1        2        2  missing
+   3 │      1        3        3  missing
+   4 │      2        4        4  missing
+   5 │      2        5        5  missing
+   6 │      2        6        6  missing
+   7 │      3        7  missing        7
+   8 │      3        8  missing        8
+   9 │      3        9  missing        9
+  10 │      4  missing  missing  missing
+
+julia> vcat(df1, df2, df3, df4, cols=:union, source=:source => 'a':'d')
+10×4 DataFrame
+ Row │ source  A        B        C
+     │ Char    Int64?   Int64?   Int64?
+─────┼───────────────────────────────────
+   1 │ a             1        1  missing
+   2 │ a             2        2  missing
+   3 │ a             3        3  missing
+   4 │ b             4        4  missing
+   5 │ b             5        5  missing
+   6 │ b             6        6  missing
+   7 │ c             7  missing        7
+   8 │ c             8  missing        8
+   9 │ c             9  missing        9
+  10 │ d       missing  missing  missing
 ```
 """
 Base.vcat(dfs::AbstractDataFrame...;
           cols::Union{Symbol, AbstractVector{Symbol},
-                      AbstractVector{<:AbstractString}}=:setequal) =
-    reduce(vcat, dfs; cols=cols)
+                      AbstractVector{<:AbstractString}}=:setequal,
+          source::Union{Nothing, SymbolOrString,
+                        Pair{<:SymbolOrString, <:AbstractVector}}=nothing) =
+    reduce(vcat, dfs; cols=cols, source=source)
 
 function Base.reduce(::typeof(vcat),
                      dfs::Union{AbstractVector{<:AbstractDataFrame},
                                 Tuple{AbstractDataFrame, Vararg{AbstractDataFrame}}};
                      cols::Union{Symbol, AbstractVector{Symbol},
-                     AbstractVector{<:AbstractString}}=:setequal)
-    return _vcat(AbstractDataFrame[df for df in dfs if ncol(df) != 0]; cols=cols)
+                                 AbstractVector{<:AbstractString}}=:setequal,
+                     source::Union{Nothing, SymbolOrString,
+                                   Pair{<:SymbolOrString, <:AbstractVector}}=nothing)
+    if source === nothing
+        return _vcat(AbstractDataFrame[df for df in dfs if ncol(df) != 0]; cols=cols)
+    else
+        len = length(dfs)
+        if source isa SymbolOrString
+            col, vals = source, 1:len
+        else
+            @assert source isa Pair{<:SymbolOrString, <:AbstractVector}
+            col, vals = source
+        end
+
+        idx = findfirst(df -> columnindex(df, col) > 0, dfs)
+        if idx !== nothing
```

Comment:
```suggestion
        if any(df -> columnindex(df, col) > 0, dfs)
```
Also would `haskey` work too? Just for the style. :-)

Reply:
1. I cannot use `any` as I need `idx` variable in the error message.
2. `AbstractDataFrame` does not support `AbstractDict` API and it does not have `haskey` method in particular